Philips.CodeAnalysis.Common/SyntaxNodeSingleDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,87 @@
+﻿// © 2023 Koninklijke Philips N.V. See License.md in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Philips.CodeAnalysis.Common
+{
+	public abstract class SingleDiagnosticAnalyzer<T> : SingleDiagnosticAnalyzer where T : SyntaxNode
+	{
+		protected SyntaxNodeAnalysisContext Context { get; private set; }
+		protected T Node { get; private set; }
+		protected string FullyQualifiedMetaDataName { get; set; }
+
+		protected SingleDiagnosticAnalyzer(DiagnosticId id, string title, string messageFormat, string description, string category,
+											Helper helper = null, DiagnosticSeverity severity = DiagnosticSeverity.Error, bool isEnabled = true)
+			: base(id, title, messageFormat, description, category, helper, severity, isEnabled)
+		{ }
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.EnableConcurrentExecution();
+
+			SyntaxKind syntaxKind = GetSyntaxKind();
+
+			if (string.IsNullOrEmpty(FullyQualifiedMetaDataName))
+			{
+				context.RegisterSyntaxNodeAction(StartAnalysis, syntaxKind);
+			}
+			else
+			{
+				context.RegisterCompilationStartAction(startContext =>
+				{
+					if (startContext.Compilation.GetTypeByMetadataName(FullyQualifiedMetaDataName) == null)
+					{
+						return;
+					}
+
+					startContext.RegisterSyntaxNodeAction(StartAnalysis, syntaxKind);
+				});
+			}
+
+		}
+
+		public void ReportDiagnostic(Location location = null, params object[] messageArgs)
+		{
+			Diagnostic diagnostic = Diagnostic.Create(Rule, location, messageArgs);
+			Context.ReportDiagnostic(diagnostic);
+		}
+
+		private void StartAnalysis(SyntaxNodeAnalysisContext context)
+		{
+			Context = context;
+			Node = (T)context.Node;
+
+			GeneratedCodeDetector generatedCodeDetector = new();
+			if (generatedCodeDetector.IsGeneratedCode(context))
+			{
+				return;
+			}
+
+			Analyze();
+		}
+
+		protected abstract void Analyze();
```

Comment:
In concurrent execution the same instance can be used in multiple thread, right? If so, it might be safer to make the `node` and `context` method parameters instead.

Reply:
Correct. However, I believe the situation is identical. These are readonly, per-instance (non-static) values, so there's no need to pass them as method parameters.

Reply:
Great catch. This caused problems (and fixed in separate PR).